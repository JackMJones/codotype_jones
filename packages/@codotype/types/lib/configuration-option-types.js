// ConfigurationOption Type IDs

const OPTION_TYPE_STRING_SELECT = 'STRING_SELECT';

// // // //

module.exports = {
  OPTION_TYPE_STRING_SELECT
}
