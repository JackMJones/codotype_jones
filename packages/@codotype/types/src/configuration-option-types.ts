export enum ConfigurationOptionsType {
  STRING_SELECT = "STRING_SELECT",
}

// // // //

// TODO - what other types need to be supported here?
// RADIO_GROUP (nice, but not necessary)
// CHECKBOXES - nice, but functionally the same as a MULTI_DROPDOWN
// DATE
// TIME
// DATETIME
// JSON - low-priority, not really necessary
export enum OptionType {
  STRING = "STRING",
  NUMBER = "NUMBER",
  BOOLEAN = "BOOLEAN",
  DROPDOWN = "DROPDOWN",
  MULTI_DROPDOWN = "MULTI_DROPDOWN",
  COLLECTION = "COLLECTION",
  INSTANCE = "INSTANCE",
}

export type OptionValue =
  | string
  | string[]
  | number
  | number[]
  | boolean
  | OptionValueInstance
  | OptionValueInstance[]
  | null;

export type OptionValueInstance =
  | OptionValueInstanceStandard
  | OptionValueInstanceAllowDisable;

// TODO - should be:
// export interface OptionValueInstance {
//     enabled: boolean;
//     value: {
//         [key: string]: OptionValue;
//     };
// }

interface OptionValueInstanceStandard {
  [key: string]: OptionValue;
}

// NOTE - this distinction is confusing
// it should always include he enabled + value pair
// When we inflate the metadata to be used in the generator, we can simplify based on the allowDisable property
interface OptionValueInstanceAllowDisable {
  enabled: boolean;
  value: {
    [key: string]: OptionValue;
  };
}

// TODO - add `description` here -> label/value might not always be enough
export interface DropdownOption {
  label: string;
  value: string;
}

// TODO - add variant to ConfigurationGroupProperty -> make list of use cases + mockups
// TODO - add `documentation` property to `ConfigurationGroupProperty` - description might not always be enough.
// TODO - add `filters` property to `ConfigurationGroupProperty` - should accept a set of ordered, pre-defined
// transformations that can be applied to a `ConfigurationGroupProperty` of a specific OptionType.
// i.e., for a Text input field, you can apply -> "strip numbers" -> "strip symbols" -> "lowercase" -> "underscored"
export interface ConfigurationGroupProperty {
  label: string;
  identifier: string;
  description: string;
  icon: string;
  type: OptionType;
  defaultValue: OptionValue;
  required: boolean;
  enabled: boolean;
  allowDisable: boolean;
  properties: ConfigurationGroupProperty[];
  dropdownOptions: DropdownOption[];
}

// TODO - integrate through the app
// TODO - add the remaining variant values here
export enum ConfigurationGroupLayoutVariant {
  list = "LIST",
}

export interface ConfigurationGroup {
  label: string;
  identifier: string;
  description: string;
  documentation: string; // Markdown
  enabled: boolean;
  allowDisable: boolean;
  variant: "LIST" | "SIDEBYSIDE"; // TODO - enum + rename!
  properties: ConfigurationGroupProperty[];
}
