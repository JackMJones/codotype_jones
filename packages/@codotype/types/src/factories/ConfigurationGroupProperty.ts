import {
  ConfigurationGroupProperty,
  DropdownOption,
  OptionType,
  OptionValue,
  PropertyLayoutVariant
} from "../configuration-option-types";
import { PropertyFilter } from "../property-filter";
import { PropertyValidation } from "../property-validation";

// // // //

interface ConfigurationGroupPropertyBuilderParams {
  id?: string;
  label: string;
  identifier: string;
  type: OptionType;
  dropdownOptions?: DropdownOption[];
  defaultValue?: OptionValue;
  description?: string;
  documentation?: string;
  icon?: string;
  required?: boolean;
  enabled?: boolean;
  allowDisable?: boolean;
  layoutVariant?: PropertyLayoutVariant;
  properties?: ConfigurationGroupProperty[];
  filters?: PropertyFilter[];
  validations?: PropertyValidation[];
}

// TODO - this should be moved into @codotype/util
// b.c this depends on `makeUniqueId` from util, but util depends on types - can't have circular dependency
export class ConfigurationGroupPropertyBuilder
  implements ConfigurationGroupProperty {
  id: string = Math.random().toString(); // TODO - replace with makeUniqueId from @codotype/util
  label: string;
  identifier: string;
  type: OptionType;
  defaultValue: OptionValue = null;
  description: string = "";
  documentation: string = "";
  icon: string = "";
  required: boolean = false; // TODO - remove this, handled by validations
  enabled: boolean = true;
  allowDisable: boolean = false; // QUESTION - should this be true?
  // layoutVariant?: PropertyLayoutVariant;
  properties: ConfigurationGroupProperty[] = [];
  dropdownOptions: DropdownOption[] = [];
  filters: PropertyFilter[] = [];
  validations: PropertyValidation[] = [];

  constructor(params: ConfigurationGroupPropertyBuilderParams) {
    this.label = params.label;
    this.identifier = params.identifier;
    this.type = params.type;

    this.id = params.id || this.id;
    this.defaultValue = params.defaultValue;
    this.description = params.description || this.description;
    this.documentation = params.documentation || this.documentation;
    this.icon = params.icon || this.icon;
    this.required = params.required || this.required;
    this.enabled = params.enabled || this.enabled;
    this.allowDisable = params.allowDisable || this.allowDisable;
    this.properties = params.properties || this.properties;
    this.dropdownOptions = params.dropdownOptions || this.dropdownOptions;
    this.filters = params.filters || this.filters;
    this.validations = params.validations || this.validations;
  }
}
