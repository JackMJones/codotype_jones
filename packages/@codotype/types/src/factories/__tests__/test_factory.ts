import { OptionType } from "../../configuration-option-types";
import { ConfigurationGroupPropertyBuilder } from "../ConfigurationGroupProperty";

// // // //

describe("factory", () => {
  test("works", () => {
    const configurationGroupProperty = new ConfigurationGroupPropertyBuilder({
      label: "meh",
      identifier: "meh_value",
      type: OptionType.BOOLEAN
    });

    expect(configurationGroupProperty).toMatchSnapshot();
  });
});
