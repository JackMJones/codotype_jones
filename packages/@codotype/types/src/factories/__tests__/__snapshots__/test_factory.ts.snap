// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`factory works 1`] = `
ConfigurationGroupPropertyBuilder {
  "allowDisable": false,
  "defaultValue": undefined,
  "description": "",
  "documentation": "",
  "dropdownOptions": Array [],
  "enabled": true,
  "filters": Array [],
  "icon": "",
  "id": "0.21643998893672078",
  "identifier": "meh_value",
  "label": "meh",
  "properties": Array [],
  "required": false,
  "type": "BOOLEAN",
  "validations": Array [],
}
`;
